로그인/로그아웃

choitaehoon opened the pull request:
- 로그인시 세션 생성, 로그아웃시 세션 소멸
- 로그인시 유저에 대한 정보가 없으면 exception handling

choitaehoon committed:
[#3] 회원가입 기능

- 회원가입 구현, 테스트 코드 작성
- 추후에 Global Exception Handling 검토

choitaehoon committed:
회원가입 리뷰 반영

- AccountController의 accountRegister api 이름 변경
- ResponseEntity 리턴 값 상수 적용
- NoArgsConstructor(AccessLevel.PROTECTED) 적용
- application.yml 로컬/운영 환경설정 분리
- 암호화 확장

choitaehoon committed:
회원가입 리뷰 반영

- ResponseEntity 리턴 값 명확한 이름으로 변경
- ResponseEntity error, success 분리(유지보수성)
- Exception Handling 적용
- EncryptionFactory bean 등록
- 추후에 sha256Util, sha256Service naming 결정

choitaehoon committed:
[#5] 로그인/로그아웃 구현

- 로그인시 세션 생성, 로그아웃시 세션 소멸
- 로그인시 유저에 대한 정보가 없으면 exception handling

f-lab-dev commented on src/main/java/com/social/instagram/controller/LoginController.java:
```diff
@@ -0,0 +1,43 @@
+
+        loginService.hasAccount(userId, password);
+
+        session.setAttribute("account", loginService.getAccount(userId));
```
`"account"`라는 세션 키가 하드코딩 되어있는데 관리의 용이성이 떨어질 것 같습니다.
로그인 관련 세션 데이터 입출력하는 책임을 클래스로 분리하여 기능을 메소드로 제공해주는건 어떨까요?

로그아웃 부분도 마찬가지입니다

choitaehoon replied:
* `"account"`를 상수화 하여 분리 하는 것이 유지보수성이 올라갈거 같네요. 
* 로그인 관련 세션 데이터 책임을 클래스로 분리하는게 저도 맞다고 생각합니다. 자신의 책임에 충실한 기능을 갖고 있으면서 필요한 작업이 생기면 작업을 수행해 달라는 요청이 객체지향적인 코드니까요. 저는 session 관련 책임을 `loginService`에 요청하고 있느니 session  클래스를 분리하여 요청하는게 맞겠네요

f-lab-dev commented on src/main/java/com/social/instagram/service/LoginService.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final AccountRepository accountRepository;
+
+    public void hasAccount(String userId, String password) {
```
이 메소드의 이름은 **해당 유저를 갖고있냐** 입니다. 그러면 기대되는 대답은 "예" 혹은 "아니오"일 것 같네요. 그럼 리턴타입은 어떻게 되야할까요? 그리고 해당 책임이 **로그인 서비스**의 책임이 맞을까요?

choitaehoon replied:
* 해당 유저를 검사하고 갖고 있지 않으면 예외를 던져서 Handler Exception이 잡는게 더 바람직하지 않을까요? 어차피 유저가 존재하지 않는다면 throw로 던져야 하니까요
* 해당 책임이 `AccountService`라고 생각이드네요. `account`에 대한 비즈니스 로직을 `LoginService`에 실행하고 있으니까요.

f-lab-dev replied:
예외를 던지는 것은 우리가 메소드를 설계할 때 의도한 동작이 아닌 예외상황에서만 던져야 합니다. 그런 의도를 갖고계시다면 그런 의도가 잘 드러나게 메소드명을 바꿔줘야겠네요.

choitaehoon replied:
`hasAccount` naming 만으로 의도한 동작을 명확하게 담지 못했네요. naming을 변경하면 `checkHasAccount`가 더 적절한 것 같네요

choitaehoon committed:
로그인/로그아웃 리뷰 반영

- session key 상수화
- 로그인 관련 세션 데이터 입출력 책임 분리

f-lab-dev commented on src/main/java/com/social/instagram/controller/LoginController.java:
```diff
@@ -13,29 +14,31 @@
+        accountService.hasAccount(userId, password);
 
-        session.setAttribute("account", loginService.getAccount(userId));
+        session.setAttribute(ACCOUNT, sessionService.getAccount(userId));
```
`sessionService`가 있음에도 해당 클래스를 거치지 않고 직접 세션에 연산하는 이유가 있으실까요?

choitaehoon replied:
생각해보니 `SessionService`의 역할인거 같네요

f-lab-dev commented on src/main/java/com/social/instagram/service/SessionService.java:
```diff
@@ -0,0 +1,17 @@
+
+    private final AccountService accountService;
+
+    public Account getAccount(String userId) {
```
이것은 session에 값을 입출력하는 책임을 가진 클래스같은데 회원과 관련된 연산을 하는 이유가 있으신가요?

choitaehoon replied:
`session`에 관한 책임인데 `Account` 책임도 더했네요. 분리하는게 맞을거 같네요

f-lab-dev commented on src/main/java/com/social/instagram/util/session/SessionNameConstants.java:
```diff
@@ -0,0 +1,7 @@
+package com.social.instagram.util.session;
+
+public class SessionNameConstants {
```
`SessionKeys` 혹은 `SessionKeyConstants` 정도로만 지어줘도 되지 않을까요? `Name`이라고 하면 어떤 역할인지 명확하지 않으니 세션관점의 용어를 써주는게 좋을 것 같아요

choitaehoon replied:
`SessionKeyConstants` naming이 괜찮아보이네요. 수정하겠습니다.

choitaehoon committed:
로그인/로그아웃 리뷰 반영

- session 값 입출력 책임 분리
- session 상수 naming 변경
- hasAccount -> checkHasAccount 메소드명 변경
- 추후에 session 리팩토링(session이 늘어남에 따라 각자에 맞는 세션 생성되므로 메소드 증가, OCP 적용 검토)

f-lab-dev commented on src/main/java/com/social/instagram/service/SessionService.java:
```diff
@@ -4,14 +4,22 @@
+    private final HttpSession session;
+
+    public void createAccountSession(Account account) {
+        session.setAttribute(ACCOUNT, account);
```
굳이 `Account` 객체를 통째로 저장할 필요가 있을까요? 세션에 저장한다면 가져다 쓸 수 있어야하는데 어떤 값은 최신성이 필요해서 세션이 아니라 DB에서 가져와야할 수도 있습니다. 혹은 어떤 데이터는 크기가 너무 클 수도 있고요. 각 데이터의 특성에 맞춰서 세션에 저장할만한것만 저장하시면 좋겠습니다

choitaehoon replied:
서비스가 확장하게 되면 'Account' 뿐만아니라 다른 데이터도 세션에 등록해 웹 서버 리소스가 부족할 수 있고 정보량이 많아 성능 또한 떨어질 수 있겠네요. 필요한 것만 세션에 등록하는게 맞겠네요.

f-lab-dev commented on src/main/java/com/social/instagram/service/AccountService.java:
```diff
@@ -25,7 +25,7 @@ public void checkDuplicateUserId(String userId) {
     }
 
-    public void hasAccount(String userId, String password) {
+    public void checkHasAccount(String userId, String password) {
```
이런것은 `validate`라는게 더 낫지 않을까요? **유효한 값이 아니면 예외를 던진다**라는 의도가 잘 나타날 것 같아요

choitaehoon replied:
`validate`가 `check` 보다 의도를 더 명확한거 같네요.

f-lab-dev commented on src/main/java/com/social/instagram/service/SessionService.java:
```diff
@@ -0,0 +1,25 @@
+
+@Service
+@RequiredArgsConstructor
+public class SessionService {
```
여기서 `SessionService`를 쓰는 이유도 주석으로 정리해주시면 좋을 것 같아요. 단순히 **책임을 나눴다** 정도가 아니라 더 자세하게요~

choitaehoon committed:
로그인/로그아웃 리뷰 반영

- 유효성, session method naming 변경
- SessionService 사용하는 이유 주석 작성
- SessionKey Account -> USER_ID naming으로 변경

f-lab-dev commented on src/main/java/com/social/instagram/service/SessionService.java:
```diff
@@ -1,25 +1,57 @@
+
+    layer별로 구성하면 안 좋은 점들이 있지만 구성하는 이유는 서비스가 확장함에 따라 장점은 더 뚜렷해지기 때문이다.
+    역할과 책임없는 코드가 흩어져서 사용되고, 불필요한 책임을 여러개 안고 있는 메소드를 테스트코드 작성한다고 하면
+    복잡도는 증가하고 가독성이 떨어지게 된다.
```
여기서는 살짝 layered architecture에 대한 강의?를 하는 것 같은데 실제 회사에서 이런 주석을 달면 이상하겠죠. Session과 관련된 연산을 하는 것을 따로 서비스로 분리한 이유. 즉 이 클래스의 존재 이유를 설명해달란 말이였습니다~

choitaehoon committed:
로그인/로그아웃 리뷰 반영

- SessionService를 분리 한 이유 주석 수정

f-lab-dev commented on src/main/java/com/social/instagram/service/SessionService.java:
```diff
@@ -8,37 +8,10 @@
+    SessionService
     객체지향적인 코드는 자신의 책임에 충실한 기능을 갖고 있으면 필요한 작업이 생기면 작업을 수행 해 달라고
     요청하는 것이다. 이를 바탕으로 코드를 보면 LoginController에게 세션을 생성하고 반환하는 것보다 LoginController에서
     SessionService에게 세션에 관한 작업을 위임하는 것이 바람직하다.
```
session을 캡슐화하여 메소드로만 기능을 제공하는 것이죠. 이런 내용도 있었으면 좋았을 것 같은데 이런 내용만 적용되면 머지하셔도 됩니다

choitaehoon committed:
로그인/로그아웃 리뷰 반영

- SessionService 캡슐화 부연설명 추가

choitaehoon committed:
Merge branch 'develop' into feature/5-login
